Test branch PR to build a working 0trace model

reethikar opened the pull request:
Barebones commit of some code that I'm experimenting with

reethikar commented:
Messing with the underlying socket on an existing websocket connection technically works but corrupts the websocket connection and closes it. For now I'm adding a second handler that speaks websocket on the server in order to do the testing. I'll update here when sending TTL limited SYN packets on the same TCP connection works right.

reethikar commented:
Based on our discussion today 7/6, I've refactored the code to send the TTL-limited packets iteratively (so incorporating a latency calculation in the future would be simpler), and now it sends a packet with TTL `t` and waits for the (ICMP) response to be received before moving on to sending a packet with TTL `t+1` and so on till `MaxTTLHops`

For now I've used a specifically crafted payload to match the ICMP time exceeded responses. My understanding of [RFC 792](https://datatracker.ietf.org/doc/html/rfc792) is that ICMP Time to live exceeded in transit response messages will contain the original datagram's data which means payloads can be used to store the identifier and then matched.

reethikar commented:
Not all hops follow the RFC792 specification, i.e. some do not send back the original datagram, so I'm using another criteria to check (could work out a different way to do the same?)

reethikar commented:
Seemingly some hops send back ICMP responses based on [RFC1812](https://datatracker.ietf.org/doc/html/rfc1812#page-53), i.e. 
> the ICMP datagram SHOULD contain as much of the original datagram as possible
   without the length of the ICMP datagram exceeding 576 bytes.  The
   returned IP header (and user data) MUST be identical to that which
   was received, except that the router is not required to undo any
   modifications to the IP header that are normally performed in
   forwarding that were performed before the error was detected

We could possible use something in the IP header (say a specifically set sequence of IPID?) to create the 1-1 mapping to measure the hops

reethikar commented:
Setting our own IPID to match packets does not work. But instead, I refactored the code to catch the sent TTL limited packets, extract the IPID `i` (say) and watch for ICMP error responses that contain the original packet's IP header with the requisite IPID `i` 

Some hops may not respond with ICMP error responses, added functionality to skip the hop after a particular amount of time `tracerouteHopTimeout`

reethikar commented:
Relevant changes are made to the `main.go` and `pingpage.html` files. 
The files in `scripts/0trace-test/` were meant to be for testing.

reethikar commented:
RTT calculation logic added now: 
I use the Timestamp embedded in the gopacket that comes from [CaptureInfo](https://pkg.go.dev/github.com/google/gopacket#CaptureInfo) whose documentation says: 
> // Timestamp is the time the packet was captured, if that is known.
	Timestamp [time](https://pkg.go.dev/time).[Time](https://pkg.go.dev/time#Time)

Comparing this timestamp of the sent packet with IP Id `i` and the corresponding ICMP error packet received that has the IP Id `i` in the original packet's IP header area gives us the RTT.

reethikar commented on cmd/test-server/main.go:
```diff
@@ -143,41 +128,271 @@ func echoHandler(w http.ResponseWriter, r *http.Request) {
+	var hopRTTVal time.Duration
+	for packet := range packetStream.Packets() {
+		// if a particular hop is taking too long, then return nil and move on to sending packets to find the next hop
+		if hasTracerouteHopTimedout(timerPerHopPerUUID[uuid]) {
+			counter += 1
+			timerPerHopPerUUID[uuid] = resetTimerForCounter()
+			// Must produce and return a <nil> value otherwise empty structs containing net.IP cannot be compared
+			var empty net.IP
+			hops <- HopRTT{empty, 0}
 		}
```
One comment/issue I have just found here: 
This goroutine only moves on to the next hop (say, `n+1`) if there is no response from a particular hop `n` and there are other packets (to/from the source) seen on the pcap handle, i.e. from the packet stream we see some packet and we realize we have been waiting beyond the `tracerouteHopTimeout` and so, move on to next hop. But, if there are no such packets seen on the handle, the goroutine will keep waiting until it sees a packet to signal this timeout. 

Is there an external timer we can create and say, look for packets **for this hop** (`for packet := range packetStream.Packets()`) only until `tracerouteHopTimeout`. If timed out, do (reset timer, increment counter, send back "empty" `HopRTT`) and then, start looking for packets for the upcoming hop.

NullHypothesis replied:
Something like the following should work:
```go
ticker := time.NewTicker(tickerFrequency)
for {
  select {
    case pkt := <- packetStream.Packets():
      // Deal with packets.
    case <-ticker.C:
      // Check if there are timeouts.
  }
}
```
The ticker fires every, say, 500 ms and allows us to check if any packets timed out.

reethikar replied:
After a bit of debugging and thinking about how to implement this logic in the recvPackets goroutine, I realized implementing it in the sending logic would make much more sense. I've done the following in b0707ed:
Send packet with TTL `t`, update a global map that keeps track of which TTL number we are currently on, then fire a ticker for `tracerouteHopTimeout`. Using the `select` goroutine wait, look for either data about the responding hop from `recvPackets` function or for the ticker to fire. If ticker fires, then enter empty hop in our result, and continue on to the next hop. 

I think this is much cleaner than trying to deal with it on the receiving side, as we will have to `.Reset(t)` the ticker each time we receive the ICMP error (that we are looking for) for the current hop.

NullHypothesis commented on cmd/test-server/main.go:
```diff
@@ -1,37 +1,49 @@
-const batchSizeLimit int = 100 // rate per batch becomes roughly 100 IPs * (5 ICMP packets + 5 TCP packets *7 ports) packets per IP
 
-var PortsToTest = [...]int{53, 80, 443, 3389, 8080, 9100}
+const batchSizeLimit int = 100
```
Nitpick: It would be more elegant to group all variables and constants together, like this:
```go
const (
  batchSizeLimit = 100
  beginTTLValue = 5
  ...
)
var (
  ...
)
```

NullHypothesis commented on cmd/test-server/main.go:
```diff
@@ -143,41 +128,271 @@ func echoHandler(w http.ResponseWriter, r *http.Request) {
-	stddev, _ := stats.StandardDeviation(data)
-	return min, avg, max, stddev
+// Parse IP and TCP headers from ICMP Response Payload
+func getHeadersFromICMPResponsePayload(data []byte) (*layers.IPv4, *layers.TCP, error) {
```
I would suggest renaming `data` to make it clear what we're dealing with. In this case, it's an ICMP packet, right? If so, maybe call it `icmpPkt` instead?

NullHypothesis commented on cmd/test-server/main.go:
```diff
@@ -143,41 +128,271 @@ func echoHandler(w http.ResponseWriter, r *http.Request) {
+	ipErr := ip.DecodeFromBytes(data[0:], gopacket.NilDecodeFeedback)
+	tcpErr := tcp.DecodeFromBytes(data[ipHeaderLength:], gopacket.NilDecodeFeedback)
+
+	if ipErr != nil && tcpErr != nil {
```
What if one variable is nil but the other isn't?

reethikar replied:
Good point, I noticed I wasn't making use of `tcp` anymore and hence decoding it (and `tcpErr`) is not useful. So I removed it.

NullHypothesis commented on cmd/test-server/main.go:
```diff
@@ -143,41 +128,271 @@ func echoHandler(w http.ResponseWriter, r *http.Request) {
+
+// Returns current time, used to reset counter timer
+func resetTimerForCounter() time.Time {
+	return time.Now()
```
I find it useful to always use UTC time. Makes it easier to deal with times if we never have to worry about time zones.
```suggestion
	return time.Now().UTC()
```

NullHypothesis commented on cmd/test-server/main.go:
```diff
@@ -143,41 +128,271 @@ func echoHandler(w http.ResponseWriter, r *http.Request) {
+}
+
+// Returns current time, used to reset counter timer
+func resetTimerForCounter() time.Time {
```
Nitpick: That's an odd name for a function that simply returns the time?

reethikar replied:
Good point, but I've done away with this function in the current version

NullHypothesis commented on cmd/test-server/main.go:
```diff
@@ -143,41 +128,271 @@ func echoHandler(w http.ResponseWriter, r *http.Request) {
+
+// Check if tracerouteHopTimeout has been reached since the timeToCheck was set
+func hasTracerouteHopTimedout(timeToCheck time.Time) bool {
+	timenow := time.Now()
```
As mentioned above, I'd use UTC.

NullHypothesis commented on cmd/test-server/main.go:
```diff
@@ -143,41 +128,271 @@ func echoHandler(w http.ResponseWriter, r *http.Request) {
+// Check if tracerouteHopTimeout has been reached since the timeToCheck was set
+func hasTracerouteHopTimedout(timeToCheck time.Time) bool {
+	timenow := time.Now()
+	if timenow.Sub(timeToCheck) >= tracerouteHopTimeout {
```
You could abbreviate those four lines to:
```go
return timenow.Sub(timeToCheck) >= tracerouteHopTimeout
```

reethikar replied:
Neat, I combined L160 into this and made it 
```
return time.Now().UTC().Sub(timeToCheck) >= tracerouteHopTimeout
```

NullHypothesis commented on cmd/test-server/main.go:
```diff
@@ -143,41 +128,271 @@ func echoHandler(w http.ResponseWriter, r *http.Request) {
+				}
+			}
+			// If it is an ICMP packet, check if it is the ICMP TTL exceeded one we are looking for
+			if icmpLayer != nil {
```
This block should probably be its own function.

NullHypothesis commented on cmd/test-server/main.go:
```diff
@@ -143,41 +128,271 @@ func echoHandler(w http.ResponseWriter, r *http.Request) {
+				icmpPkt, _ := icmpLayer.(*layers.ICMPv4)
+				ipHeaderIcmp, _, err := getHeadersFromICMPResponsePayload(icmpPkt.LayerPayload())
+				if err != nil {
+					// ErrLogger.Println("Error getting header from ICMP packet: ", err)
```
Nitpick: Source code that's commented out should not be in version control.

NullHypothesis commented on cmd/test-server/main.go:
```diff
@@ -143,41 +128,271 @@ func echoHandler(w http.ResponseWriter, r *http.Request) {
+			ipl, _ := ipLayer.(*layers.IPv4)
+			currHop := ipl.SrcIP.String()
+			// To identify the TCP packet that we have sent
+			if tcpLayer != nil {
```
This block can be its own function.

NullHypothesis commented on cmd/test-server/main.go:
```diff
@@ -143,41 +128,271 @@ func echoHandler(w http.ResponseWriter, r *http.Request) {
-	if err == nil || strings.Contains(err.Error(), "connection refused") {
-		if err == nil {
-			defer conn.Close()
+// Listen on the provided pcap handler for packets sent
```
[Godoc](https://go.dev/doc/comment) requires that function comments start with the name of the function, i.e.:
```go
// recvPacket does x, y, and z.
func recvPackets(...) {
  ...
```
For complex functions whose parameters aren't immediately obvious, I suggest documenting the parameters in the function comment.

NullHypothesis commented on cmd/test-server/main.go:
```diff
@@ -143,41 +128,271 @@ func echoHandler(w http.ResponseWriter, r *http.Request) {
+	)
+	if serializeErr != nil {
+		ErrLogger.Println("Serialize error: ", serializeErr)
+		return false
```
I assume the boolean signals to the caller that something went wrong? If so, I would return an `error` instead.

NullHypothesis commented on cmd/test-server/main.go:
```diff
@@ -143,41 +128,271 @@ func echoHandler(w http.ResponseWriter, r *http.Request) {
+
+// Reach the underlying connection and set up necessary handler and initalize 0trace set up
+func start0trace(uuid string, clientIP string, clientPort string, netConn net.Conn) map[int]HopRTT {
+	handle, err := pcap.OpenLive(deviceName, 65536, true, time.Second)
```
Literals make code unnecessarily hard to read. I would turn 65536 into a constant, e.g., `const snapLen = 65536` and reference it here.

NullHypothesis commented on cmd/test-server/main.go:
```diff
@@ -143,41 +128,271 @@ func echoHandler(w http.ResponseWriter, r *http.Request) {
+}
+
+// Reach the underlying connection and set up necessary handler and initalize 0trace set up
+func start0trace(uuid string, clientIP string, clientPort string, netConn net.Conn) map[int]HopRTT {
```
I would convert `clientPort` to an integer in the calling function. It's a bit odd for a function to accept a port as string.

NullHypothesis commented on cmd/test-server/main.go:
```diff
@@ -143,41 +128,271 @@ func echoHandler(w http.ResponseWriter, r *http.Request) {
+		hopData := <-recvdHop
+		traceroute[ttlValue] = hopData
+		if hopData.IP == nil {
+			ErrLogger.Println("Moving on to the next hop")
```
Looks like there's a `continue` missing here?

reethikar replied:
Right, I should have an explicit continue instead of letting it (uselessly) execute the rest of the body

NullHypothesis commented on cmd/test-server/main.go:
```diff
@@ -143,41 +128,271 @@ func echoHandler(w http.ResponseWriter, r *http.Request) {
+	}
+	defer c.Close()
+	myConn := c.UnderlyingConn()
+	traceroute := start0trace(uuid, clientIP, clientPort, myConn)
```
Doesn't `myConn` already contain the client's address and port? Via `myConn.RemoteAddr()`? If so, `clientIP` and `clientPort` are redundant.

NullHypothesis commented on cmd/test-server/main.go:
```diff
@@ -143,41 +128,271 @@ func echoHandler(w http.ResponseWriter, r *http.Request) {
+
+	c, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		ErrLogger.Println("upgrade:", err)
```
Don't we want to inform the client about this error? E.g., by doing:
```go
http.Error(w, err.Error(), http.StatusInternalServerError)
```

NullHypothesis commented on cmd/test-server/main.go:
```diff
@@ -143,41 +128,271 @@ func echoHandler(w http.ResponseWriter, r *http.Request) {
+		Timestamp: time.Now().UTC().Format("2006-01-02T15:04:05.000000"),
+		HopData:   traceroute,
+	}
+	zeroTraceResult, _ := json.Marshal(results)
```
Same as above: We may want to check for an error and inform the client.

reethikar replied:
We don't display this back to the user, but may be logging something is important.

NullHypothesis commented on cmd/test-server/pingpage.html:
```diff
@@ -81,13 +82,36 @@ <h2>Calculate Latency Using WebSockets and Measure ICMP and TCP latency</h2>
+          const socket = new WebSocket(webSocketAddr);
+          var startTime = new Date().getTime(); 
+          const intervalId = setInterval(() => {
+            if(new Date().getTime() - startTime > 600000){
```
Same as above: Use constants instead of int literals.

NullHypothesis commented on cmd/test-server/pingpage.html:
```diff
@@ -81,13 +82,36 @@ <h2>Calculate Latency Using WebSockets and Measure ICMP and TCP latency</h2>
+              return;
+            }
+            socket.send("Keep alive");
+          }, 1000);
```
Same as above: Use constants instead of int literals.

NullHypothesis commented on scripts/0trace-test/modify-websocket.go:
```diff
@@ -0,0 +1,185 @@
+	"time"
+)
+
+var addr = flag.String("addr", ":80", "http service address")
```
Nitpick: There does not seem to be a need for this variable to be global.

reethikar commented:
Thanks for the review, and I've addressed the PR comments necessary and fixed issues (including #4). I'll mark it for review again as I've changed things considerably (i.e. b0707ed).

NullHypothesis commented on scripts/0trace-test/modify-websocket.go:
```diff
@@ -0,0 +1,184 @@
+
+var upgrader = websocket.Upgrader{} // use default options
+
+func tcpConn(clientIP string, clientPort string, netConn net.Conn) {
```
The name's a bit vague. How about `sendTcpSeg` instead?

NullHypothesis commented on scripts/0trace-test/modify-websocket.go:
```diff
@@ -0,0 +1,184 @@
+
+	n, err := tcpConn.Write([]byte(outgoingPacket))
+	if err != nil {
+		log.Print(err)
```
This error seems worth propagating to me, i.e., return it back to the caller, which is the echo handler. Also, if we fail to write the packets to the wire, is there even a point in continuing execution in the echo handler?
